From 48d51ea72f22f7e08aece7fd3b0140258ad35d86 Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Thu, 28 Jul 2022 05:11:12 +0000
Subject: [PATCH] xen/arm: mpam: Enable MPAM at boot time

This commit adds the enabling of MPAM once all MSC devices are probed
and initialized. A global variable `mpam_enabled` is introduced as
the flag indicating if the MPAM has been enabled in the system.

Since function `mpam_enable()` is supposed to be called as the
last step of `mpam_discovery_cpu_online()` (which happens on every
pCPU), use the global variable `mpam_enabled` in the early stage
of `mpam_discovery_cpu_online()` so that unnecessary MSC probing
stages can be skipped.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-5049
Change-Id: Ib3996a28c5df176fef1fba6214119f0393f50504
---
 xen/arch/arm/include/asm/mpam.h |  1 +
 xen/arch/arm/mpam.c             | 44 +++++++++++++++++++++++++++++++++
 xen/arch/arm/setup.c            |  4 +++
 3 files changed, 49 insertions(+)

diff --git a/xen/arch/arm/include/asm/mpam.h b/xen/arch/arm/include/asm/mpam.h
index d4454005b3..9c765f6068 100644
--- a/xen/arch/arm/include/asm/mpam.h
+++ b/xen/arch/arm/include/asm/mpam.h
@@ -393,6 +393,7 @@ struct mpam_msc_ris
     struct mpam_component *comp;
 };
 
+extern void mpam_postsmp_setup(void);
 extern void mpam_presmp_setup(void);
 extern int mpam_discovery_cpu_online(unsigned int cpu);
 
diff --git a/xen/arch/arm/mpam.c b/xen/arch/arm/mpam.c
index 72d76ec900..b2a031d0c3 100644
--- a/xen/arch/arm/mpam.c
+++ b/xen/arch/arm/mpam.c
@@ -51,6 +51,8 @@ static unsigned int mpam_fw_num_msc;
 static uint16_t mpam_partid_max;
 static uint8_t mpam_pmg_max;
 
+static bool __read_mostly mpam_enabled = false;
+
 static struct dt_device_node *
 __init dt_find_next_cache_node(const struct dt_device_node *np)
 {
@@ -1004,6 +1006,42 @@ static void __init mpam_cpu_online(unsigned int cpu)
     spin_unlock(&mpam_list_lock);
 }
 
+/* Enable MPAM once all devices have been probed. */
+static int __init mpam_enable(void)
+{
+    struct mpam_msc *msc;
+    bool all_devices_probed = true;
+
+    spin_lock(&mpam_list_lock);
+    /* Have we probed all the devices? */
+    list_for_each_entry(msc, &mpam_all_msc, glbl_list)
+    {
+        spin_lock(&msc->lock);
+        if ( !msc->probed )
+            all_devices_probed = false;
+        spin_unlock(&msc->lock);
+
+        if ( !all_devices_probed )
+            break;
+    }
+
+    if ( !all_devices_probed )
+    {
+        spin_unlock(&mpam_list_lock);
+        return -ENODEV;
+    }
+
+    /* With all MSCs probed, now we can set the global variable to true. */
+    mpam_enabled = true;
+
+    printk(XENLOG_INFO "MPAM: MPAM enabled with %u partid(s) and %u pmg(s).\n",
+           ACCESS_ONCE(mpam_partid_max) + 1, ACCESS_ONCE(mpam_pmg_max) + 1);
+
+    spin_unlock(&mpam_list_lock);
+
+    return 0;
+}
+
 static int __init mpam_msc_probe(void)
 {
     unsigned int num_msc = 0;
@@ -1054,6 +1092,12 @@ void __init mpam_presmp_setup(void)
                "MPAM: Error in probing MPAM hardware when CPU0 is online\n");
 }
 
+void __init mpam_postsmp_setup(void)
+{
+    if ( mpam_enable() )
+        printk(XENLOG_ERR "MPAM: Unable to enable MPAM\n");
+}
+
 int __init mpam_discovery_cpu_online(unsigned int cpu)
 {
     int err = 0;
diff --git a/xen/arch/arm/setup.c b/xen/arch/arm/setup.c
index b17b128191..bdcbd01853 100644
--- a/xen/arch/arm/setup.c
+++ b/xen/arch/arm/setup.c
@@ -1141,6 +1141,10 @@ void __init start_xen(unsigned long boot_phys_offset,
     printk("Brought up %ld CPUs\n", (long)num_online_cpus());
     /* TODO: smp_cpus_done(); */
 
+#ifdef CONFIG_MPAM
+    mpam_postsmp_setup();
+#endif
+
     /* This should be done in a vpmu driver but we do not have one yet. */
     vpmu_is_available = cpu_has_pmu;
 
-- 
2.25.1

